File: RayTracerCSSingleProj/Colour.cs
using System;

namespace RayTracer
{
public class Colour
{
    private double R, G, B; 

    public Colour(double r, double g, double b) {
        R = r;
        G = g;
        B = b;
    }

        public Colour(string str)
        {
            string[] nums = str.Split(',');
            if (nums.Length != 3) throw new ArgumentException();
            R = double.Parse(nums[0]);
            G = double.Parse(nums[1]);
            B = double.Parse(nums[2]);
        }

        public Colour MultiplyBy(double n)
        {
            return new Colour(n * R, n * G, n * B);
        }
        public Colour MultiplyBy(Colour c2)
        {
            return new Colour(R * c2.R, G * c2.G, B * c2.B);
        }
        public Colour Plus(Colour c2)
        {
            return new Colour(R + c2.R, G + c2.G, B + c2.B);
        }
        public Colour Minus(Colour c2)
        {
            return new Colour(R - c2.R, G - c2.G, B - c2.B);
        }

        private byte ToByte(double d)
        {
            return Convert.ToByte(Math.Min(1,d)*255);
        }

        public byte RedByte()
        {
            return ToByte(R);
        }

        public byte GreenByte()
        {
            return ToByte(G);
        }

        public byte BlueByte()
        {
            return ToByte(B);
        }

        public static readonly Colour Red = new Colour(1, 0, 0);
        public static readonly Colour Green = new Colour(0, 1, 0);
        public static readonly Colour Blue = new Colour(0, 0, 1);
        public static readonly Colour Yellow = new Colour(1, 1, 0);
        public static readonly Colour White = new Colour(1, 1, 1);
        public static readonly Colour Black = new Colour(0, 0, 0);

        #region Operators
        public static Colour operator *(double n, Colour c1)
        {
            return c1.MultiplyBy(n);
        }
        public static Colour operator *(Colour c1, Colour c2)
        {
            return c1.MultiplyBy(c2);
        }
        public static Colour operator +(Colour c1, Colour c2)
        {
            return c1.Plus(c2);
        }
        public static Colour operator -(Colour c1, Colour c2)
        {
            return c1.Minus(c2);
        }
        #endregion
    }
}

File: RayTracerVBSingleProj/Colour.vb
Public Class Colour
    Private R, G, B As Double

    Public Sub New(ByVal r As Double, ByVal g As Double, ByVal b As Double)
        Me.R = r
        Me.G = g
        Me.B = b
    End Sub

    Public Sub New(ByVal str As String)
        Dim nums As String() = str.Split(","c)
        If nums.Length <> 3 Then Throw New ArgumentException()
        R = Double.Parse(nums(0))
        G = Double.Parse(nums(1))
        B = Double.Parse(nums(2))
    End Sub

    Public Function MultiplyBy(ByVal n As Double) As Colour
        Return New Colour(n * R, n * G, n * B)
    End Function

    Public Function MultiplyBy(ByVal c2 As Colour) As Colour
        Return New Colour(R * c2.R, G * c2.G, B * c2.B)
    End Function

    Public Function Plus(ByVal c2 As Colour) As Colour
        Return New Colour(R + c2.R, G + c2.G, B + c2.B)
    End Function

    Public Function Minus(ByVal c2 As Colour) As Colour
        Return New Colour(R - c2.R, G - c2.G, B - c2.B)
    End Function

    Private Function ToByte(ByVal d As Double) As Byte
        Return Convert.ToByte(Math.Min(1, d) * 255)
    End Function

    Public Function RedByte() As Byte
        Return ToByte(R)
    End Function

    Public Function GreenByte() As Byte
        Return ToByte(G)
    End Function

    Public Function BlueByte() As Byte
        Return ToByte(B)
    End Function

    Public Shared Red As Colour = New Colour(1, 0, 0)
    Public Shared Green As Colour = New Colour(0, 1, 0)
    Public Shared Blue As Colour = New Colour(0, 0, 1)
    Public Shared Yellow As Colour = New Colour(1, 1, 0)
    Public Shared White As Colour = New Colour(1, 1, 1)
    Public Shared Black As Colour = New Colour(0, 0, 0)

#Region "Operators"

    Public Shared Operator *(ByVal n As Double, ByVal c1 As Colour) As Colour
        Return c1.MultiplyBy(n)
    End Operator

    Public Shared Operator *(ByVal c1 As Colour, ByVal c2 As Colour) As Colour
        Return c1.MultiplyBy(c2)
    End Operator

    Public Shared Operator +(ByVal c1 As Colour, ByVal c2 As Colour) As Colour
        Return c1.Plus(c2)
    End Operator

    Public Shared Operator -(ByVal c1 As Colour, ByVal c2 As Colour) As Colour
        Return c1.Minus(c2)
    End Operator
#End Region
End Class


